### PR TITLE
ptrace(2) tests and fixes

### PR DESCRIPTION
**Pull request by bsdjhb**

### CODE REVIEW

File: sys/kern/sys_process.c [hunk 1]
```diff
@@ -529,11 +529,11 @@ proc_read_cheri_tags(struct proc *p, struct uio *uio)
 	 * based on the tag bitmask size.
 	 */
 	va = uio->uio_offset;
-	if (!is_aligned(va, sizeof(uintcap_t)))
+	if (!is_aligned(va, sizeof(uintcap_t) * 8))
```
Comment (gvnn3) [inline sys/kern/sys_process.c:532]: Can I ask that 8 be #defined to a name?

**Comment by bsdjhb** [comment]
Humm, the test failures are odd.  I don't understand why PT_ATTACH would be failing.  It doesn't seem deterministic.  I wonder if it's some kind of race between PT_TRACE_ME in the child and PT_ATTACH in the parent?  The ptrace tests in upstream use this approach without issue.

**Comment by bsdjhb** [comment]
I had mis-read the upstream tests.  They don't use PT_ATTACH at all when the child does PT_TRACE_ME.

File: sys/kern/sys_process.c [hunk 2]
```diff
@@ -529,11 +529,11 @@ proc_read_cheri_tags(struct proc *p, struct uio *uio)
 	 * based on the tag bitmask size.
 	 */
 	va = uio->uio_offset;
-	if (!is_aligned(va, sizeof(uintcap_t)))
+	if (!is_aligned(va, sizeof(uintcap_t) * CHAR_BIT))
```
Comment (arichardson) [inline sys/kern/sys_process.c:532]: Why is this change needed? The one below makes sense to me, but doesn't this restrict the VA to be a multiple of 128 bytes?

File: bin/cheribsdtest/cheribsdtest_ptrace.c [hunk 3]
```diff
@@ -156,3 +156,60 @@ CHERIBSDTEST(test_ptrace_readtags, "Basic test of PIOD_READ_CHERI_TAGS")
...
+	fpid = fork();
+	CHERIBSDTEST_VERIFY2(fpid != -1, "Could not fork: errno=%d", errno);
+
+	if (fpid == 0) {
+		/* child */
+		CHERIBSDTEST_VERIFY(ptrace(PT_TRACE_ME, 0, NULL, 0) == 0);
+		raise(SIGSTOP);
+
+		exit(0);
+	}
```
Comment (arichardson) [inline bin/cheribsdtest/cheribsdtest_ptrace.c:188]: This code is duplicated three times now, could factor it out into a helper?

File: bin/cheribsdtest/cheribsdtest_ptrace.c [hunk 4]
```diff
@@ -156,3 +156,60 @@ CHERIBSDTEST(test_ptrace_readtags, "Basic test of PIOD_READ_CHERI_TAGS")
...
+	piod.piod_offs = &pp[last_index];
+	piod.piod_addr = capbuf;
+	piod.piod_len = sizeof(capbuf);
+	CHERIBSDTEST_VERIFY(ptrace(PT_IO, wpid, (caddr_t)&piod, 0) == 0);
+
+	CHERIBSDTEST_VERIFY(piod.piod_len == sizeof(capbuf));
+	CHERIBSDTEST_VERIFY2(capbuf[0] == 1,
+	    "Tag not set in returned buffer");
+	memcpy(&cap, &capbuf[1], sizeof(cap));
+	CHERIBSDTEST_VERIFY2(cheri_cleartag(pp[last_index]) == cap,
```
Comment (arichardson) [inline bin/cheribsdtest/cheribsdtest_ptrace.c:205]: Not comparing all bits.

File: sys/kern/sys_process.c (hunk 2 above)
Comment (bsdjhb) [reply sys/kern/sys_process.c:532]: Otherwise I need to do some bit-shifting nonsense to deal with misaligned tag fetches.  For example, if you fetch 8 tags starting at the second capability in a page, then I have to take the first two bytes of the tags returned from the page and synthesize the returned 8 bits by selecting 7 bits from the first byte and 1 bit from the second byte.  I didn't feel like solving that.

Comment (arichardson) [reply sys/kern/sys_process.c:532]: Ah yes that makes sense. Might be worth adding a comment that the restriction is there for simplicity of implementation rather than fundamental.

**Review by arichardson — APPROVED** [review]

File: bin/cheribsdtest/cheribsdtest_ptrace.c [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+#include <cheri/cheric.h>
+
+#include "cheribsdtest.h"
+
+/*
+  * Fork a child process and attach to it in the parent.  The child
+  * exits, but the parent returns the child pid to the caller after
+  * attaching to the stopped child.
+  */
```
Comment (jrtc27) [inline bin/cheribsdtest/cheribsdtest_ptrace.c:53]:
```suggestion
 * Fork a child process and attach to it in the parent.  The child
 * exits, but the parent returns the child pid to the caller after
 * attaching to the stopped child.
 */
```
